### PR TITLE
Unified iterator types [ECR-4192]

### PR DESCRIPTION
**Pull request by slowli**
## Overview
 
The season of last-minute breaking changes continues! This time, it's unification of iterator types for all MerkleDB indexes. This is (hopefully) better for bindings (they can now deal with a single iterator type extracted via a unified interface for all indexes), and simplifies support.

There are also some moderately unrelated changes:

- `KeySetIndex::insert` now takes the element by reference.
- `KeySetIndex` now supports unsized keys and does not have `K: Borrow<Q>` shenanigans going on, similar to `MapIndex` / `ProofMapIndex`
- Maximum height in `ProofListIndex` was fixed from the bogus value 58 to 56 (needs more detailed testing; I'm not sure this is fixed *everywhere*, and that there are no off-by-one errors)
- Bogus setting of the empty key was removed for `ListIndex`. (This code is seemingly an artifact from an era where indexes did not have metadata.)

Discussion:

- Does this really help bindings?
- Does iteration with `Key = ()`  need to be implemented for `Entry` / `ProofEntry` for uniformity?

------

See: https://jira.bf.local/browse/ECR-4192

### CODE REVIEW

File: components/merkledb/src/indexes/value_set.rs [hunk 1]
```diff
@@ -134,7 +107,8 @@ where
         self.base.contains(hash)
     }
 
-    /// Returns an iterator visiting all elements in arbitrary order. The iterator element type is V.
+    /// Returns an iterator over set elements and their hashes. The elements are ordered as per
+    /// lexicorgraphic ordering of their hashes (i.e., effectively randomly).
```
Comment (aleksuss) [inline components/merkledb/src/indexes/value_set.rs]:
```suggestion
    /// lexicographic ordering of their hashes (i.e., effectively randomly).
```

File: components/merkledb/src/indexes/iter.rs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+};
+
+/// Iterator over key-value pairs of an index.
+///
+/// This structure is returned by the [`IndexIterator`] trait and by inherent methods
+/// of some indexes.
+///
+/// [`IndexIterator`]: trait.IndexIterator.html
+#[derive(Debug)]
+pub struct Entries<'a, K: ?Sized, V> {
```
Comment (aleksuss) [inline components/merkledb/src/indexes/iter.rs]: The plural is a little bit confusing here 🤔

Comment (slowli) [reply components/merkledb/src/indexes/iter.rs]: But this is how iterators are usually named in the standard Rust library. Although, in this particular case, `Iter` is another option.
